File: Lucy/2025/4-W02/파괴되지_않은_건물.js
function solution(board, skill) {
  var answer = 0;

  const TYPE = {
    enemy: 1,
    friendly: 2,
  };

  const ROW = board.length;
  const COL = board[0].length;
  const imosBoard = Array.from({ length: ROW + 1 }, () => new Array(COL + 1).fill(0));

  for (const [type, startRow, startCol, endRow, endCol, degree] of skill) {
    let val = 1;
    if (type === TYPE.enemy) val = -1;
    else val = 1;

    imosBoard[startRow][startCol] += degree * val;
    imosBoard[startRow][endCol + 1] -= degree * val;
    imosBoard[endRow + 1][startCol] -= degree * val;
    imosBoard[endRow + 1][endCol + 1] += degree * val;
  }

  // 행 누적합
  for (let i = 0; i < ROW + 1; i++) {
    for (let j = 1; j < COL + 1; j++) {
      imosBoard[i][j] += imosBoard[i][j - 1];
    }
  }

  // 열 누적합
  for (let j = 0; j < COL + 1; j++) {
    for (let i = 1; i < ROW + 1; i++) {
      imosBoard[i][j] += imosBoard[i - 1][j];
    }
  }

  for (let i = 0; i < ROW; i++) {
    for (let j = 0; j < COL; j++) {
      const durability = board[i][j] + imosBoard[i][j];
      if (durability > 0) answer++;
    }
  }

  return answer;
}

solution(
  [
    [5, 5, 5, 5, 5],
    [5, 5, 5, 5, 5],
    [5, 5, 5, 5, 5],
    [5, 5, 5, 5, 5],
  ],
  [
    [1, 0, 0, 3, 4, 4],
    [1, 2, 0, 2, 3, 2],
    [2, 1, 0, 3, 1, 2],
    [1, 0, 1, 3, 3, 1],
  ]
); // 10

File: Gombang/Programmers/파괴되지 않은 건물.cpp
#include <string>
#include <vector>

using namespace std;

int sumArray[1010][1010];
int N, M;

int solution(vector<vector<int>> board, vector<vector<int>> skill) {
    N = board.size(), M = board[0].size();

    for (auto sk : skill) {
        int degree = sk[5];
        if (sk[0] == 1) degree = -degree;

        sumArray[sk[1]][sk[2]] += degree;
        sumArray[sk[3] + 1][sk[4] + 1] += degree;
        sumArray[sk[1]][sk[4] + 1] -= degree;
        sumArray[sk[3] + 1][sk[2]] -= degree;
    }

    // 왼쪽에서 오른쪽으로 누적합을 구하는 과정.
    for (int i = 0; i < N + 1; ++i) {
        for (int j = 0; j < M; ++j) {
            sumArray[i][j + 1] += sumArray[i][j];
        }
    }

    // 위에서 아래로 누적합을 구하는 과정.
    for (int j = 0; j < M + 1; ++j) { 
        for (int i = 0; i < N; ++i) {
            sumArray[i + 1][j] += sumArray[i][j];
        }
    }

    // 최종적으로 보드에 대해 sumArray적용.
    int answer = 0;
    for (int i = 0; i < N; ++i) { 
        for (int j = 0; j < M; ++j) {
            if (board[i][j] + sumArray[i][j] > 0) answer++;
        }
    }

    return answer;
}